### DISCUSSION TITLE
Multiple patch operations impacts read performance

### DISCUSSION THREAD

**Discussion by CyberCyclone** [ravendb/ravendb#13081] [Q&A] [answered]
Hi,

I have a bunch of patch operations like the below operating on about 1,300 documents:

```
await ravenSession.advanced.patch<StakePool, any>(`stakePool/${RawStakePool.hash}`, 'lastModified', lastModified);
```
There's 10 of the above for 1 session and then saved. 

The above session operations run in parallel totalling 50 tasks and RavenDB dashboard says there's about 12 writes per second.

While the above is running, I tried running the below query:

```
const rawResult: any[] = await session.query({collection: "StakePools"})
        .statistics(s => { stats = s; })
        .selectFields([ 
			"id",
			"margin",
			"pledge",
			"retirements",
			"meta",
			"promo",
			"blocks",
			"updatedIn",
			"activeStake",
			"delegators",
			"saturation",
			"relays",
			"url",
			"tags"
		 ])
        .all();
```

While the 50 sessions of patch operations are running, the above query can take from 8 seconds to 45 seconds to return. I've tried querying against the collection and an index and it still behaves the same. When I stop the patch operations the query returns nearly instantly.

Is there any obvious reason why such a basic query would be slowed down by the patch operations on such a small dataset? The servers I'm running this on are Hetzner cloud servers, with 2 cores assigned (both the server and RavenDB) and the read operations are against a secondary server. Hetzner servers have high IOPS and there's no CPU load, although memory is a bit tight at 2GB.

Thanks!
Casey.

**Comment by ayende** [comment]
That is a shock to me. I can't think of any single reason why that would be the case.

Patching operations require us to load, modify and persist the document, sure, but that is done in the transaction merger thread.
Both indexing and queries are actually running on separate threads and there are no locks shared between them.

How much total data do you have?
What does the Traffic Watch says? 

Is this something that you can reproduce in isolate code? 
The http://live-test.ravendb.net/ instance is a far weaker machine than your specs (2 cores, 512 MB, low I/O, etc).

I run the following code on that:

```
const RavenDocumentStore = require("ravendb");
var store  = new RavenDocumentStore.DocumentStore('http://live-test.ravendb.net/', 'ggg');
store.initialize();

class StakePool
{
    constructor() {

        this.lastModified = new Date();
    }
}

async function query(){
    for (let i = 0; i < 50000; i++) {
        const ravenSession = store.openSession();
        var start = new Date()
        const rawResult  = await ravenSession.query({collection: "StakePools"})
            .all();
        var end = new Date() - start
        console.log("query took: " + end);
    }
}

async function populate(){
    if(false) // first time only...
    {
        const ravenSession = store.openSession();

        for(let i = 0 ; i < 100 ; i++){
            await ravenSession.store(new StakePool(), "StakePools/" + i );
        }
        await ravenSession.saveChanges();    
    }
    console.log("starting patching");
    for (let i = 0; i < 500; i++) {
        const ravenSession = store.openSession();
        var base = Math.floor(Math.random() * 100);
        for(let x = 0 ; x < 10; x++){
            await ravenSession.advanced.patch(`stakePool/${(x + base) % 100}`, 'lastModified', new Date());
        }
        await ravenSession.saveChanges();
    }
}
query();
setTimeout(populate, 5000);
```

The output was:

```
query took: 246
query took: 242
query took: 243
query took: 244
starting patching
query took: 244
query took: 242
query took: 244
query took: 244
query took: 246
```

The time it takes didn't change in any way, the actual cost is mostly the network latency.

Can you look at the code and try to reproduce the issue?

**Reply by CyberCyclone** [reply]
Awesome, if it's not a known issue then it sounds like something with my setup. I'll try and replicate it and get back to you in the next day or two with the results.

**Comment by CyberCyclone** [comment]
Disregard - Please see below seperate answer.

--------------------------------

Hi @ayende, thanks for the time you put into creating the test script. I adapted it to my use case and using the "chance" npm library to keep it as similar as possible to my dataset.

From what I can see, the `patch` statements increase the latency of the query returning however the query itself returns reasonably quickly.

With the bulk statements running, it returns in 18 seconds (query time of 651ms).
```
stats QueryStatistics {
  isStale: false,
  durationInMs: 651,
  totalResults: 1300,
  skippedResults: 0,
  timestamp: 0000-12-31T13:47:52.000Z,
  indexName: 'collection/StakePoolTests',
  indexTimestamp: 0000-12-31T13:47:52.000Z,
  lastQueryTime: 0000-12-31T13:47:52.000Z,
  resultEtag: -5038121197459177000,
  nodeTag: 'A'
}
query took: 18917
```

Once the bulk statements finish, the queries finish in about 4 seconds (queries are cached).
```
stats QueryStatistics {
  isStale: false,
  durationInMs: -1,
  totalResults: 1300,
  skippedResults: 0,
  timestamp: 0000-12-31T13:47:52.000Z,
  indexName: 'collection/StakePoolTests',
  indexTimestamp: 0000-12-31T13:47:52.000Z,
  lastQueryTime: 0000-12-31T13:47:52.000Z,
  resultEtag: 167283547882215140,
  nodeTag: 'A'
}
query took: 4843
```


The updated script:
1. `if` statement in `populate` function needs to be set to 'true' for first runs
2. The script will run `Doing bulk patch` a dozen times and then stop and just run the queries.
```
const RavenDocumentStore = require("ravendb");
const Chance = require("chance");
var chance = new Chance();

var store  = new RavenDocumentStore.DocumentStore('http://live-test.ravendb.net/', 'StakePoolData');
store.initialize();

class StakePoolTest
{
    constructor() {
        
        this.lastModified = new Date(),
        this.pledge = chance.integer({ min: 100000000000, max: 1000000000000 }),
        this.poolRegistrationTimestamp = chance.date(),
        this.meta = {
            "name": chance.name(),
            "description": chance.sentence(),
            "ticker": chance.word().toUpperCase(),
            "homepage": chance.domain(),
            "extended": chance.domain(),
            "lastRetrieved": chance.date(),
            "statusCode": chance.integer({ min: 200, max: 400 })
        },
        this.relays = randomRelays(),
        this.activeStake = chance.integer({ min: 1, max: 20000000000000 }),
        this.saturation = chance.integer({ min: 1, max: 100 }),
        this.delegators = chance.integer({ min: 1, max: 1000 }),
        this.blocks = chance.integer({ min: 1, max: 10000 }),
        this.retirements = [],
        this.updatedIn = {
            "includedAt": chance.date(),
            "block": {
                "number": chance.integer({ min: 1, max: 3000000 })
            }
        }
    }
}

async function query(){
    /**
     *   !!!!!!!!!!!!! We need this to run until script is forcefully stopped. 
     * */
    while (true) {
        const ravenSession = store.openSession();
        var start = new Date()
        let stats;
        const rawResult = await ravenSession.query({collection: "StakePoolTests"})
            .statistics(s => { stats = s; })
            .selectFields([ 
                "id",
                "margin",
                "pledge",
                "retirements",
                "meta",
                "promo",
                "blocks",
                "updatedIn",
                "activeStake",
                "delegators",
                "saturation",
                "relays",
                "url",
                "tags",
                "poolRegistrationTimestamp"
             ])
            .all();
        console.log('stats', stats);

        var end = new Date() - start
        console.log("query took: " + end);
    }
}

function randomRelays(){
    let relays = [];
    for(let i = 0; i < 10; i++){
        let temp = {
            "ipv4": chance.ip(),
            "ipv6": chance.ipv6(),
            "dnsName": chance.domain(),
            "dnsSrvName": null,
            "port": chance.integer({ min: 3000, max: 9000 }),
            "alive": chance.bool(),
            "maxmind": {
                "city": {
                    "names": {
                        "en": chance.city(),
                    }
                },
                "country": {
                    "iso_code":  chance.country( ),
                },
                "location": {
                    "accuracy_radius": chance.integer({ min: 1, max: 100 }) ,
                    "latitude": chance.longitude(),
                    "longitude": chance.longitude(),
                    "time_zone": chance.timezone().utc
                },
                "autonomous_system_organization": chance.name()
            },
            "lastTested": chance.date(),
            "port22": chance.pickone(["closed", "open"])
        }
        relays.push(temp);
    };
    return relays;
}

async function populate(){
    if(false) // first time only...
    {
        const ravenSession = store.openSession();

        for(let i = 0 ; i < 1300 ; i++){
            await ravenSession.store(new StakePoolTest(), "StakePoolTests/" + i );
        }
        await ravenSession.saveChanges();    
    }
    console.log("starting patching");
    let promiseArr = [];
    for (let i = 0; i < 1300; i++) {
        // Only process 50 at a time.
        if (i && i % 50 === 0) {
            // We have 50 Promises, now run them and wait.
            console.log('Doing bulk patch');
            await Promise.all(promiseArr);
            // Done, clear the promises for the next batch.
            promiseArr = [];
        }
        let promiseHolder = new Promise(async function(resolve, reject){
            const ravenSession = store.openSession();
            let randomId = chance.integer({ min: 1, max: 1300 })
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'pledge', chance.integer({ min: 100000000000, max: 1000000000000 }));
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'updatedIn', chance.date());
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'lastModified', new Date());
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'delegators', chance.integer({ min: 1, max: 10000 }));
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'poolRegistrationTimestamp', chance.date());
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'blocks', chance.integer({ min: 1, max: 10000 }));
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'activeStake', chance.integer({ min: 1, max: 200000000 }));
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'saturation', chance.integer({ min: 1, max: 100 }));
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'meta', {
                "name": chance.name(),
                "description": chance.sentence(),
                "ticker": chance.word().toUpperCase(),
                "homepage": chance.domain(),
                "extended": chance.domain(),
                "lastRetrieved": chance.date(),
                "statusCode": chance.integer({ min: 200, max: 400 })
            });
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'meta.extended', {
                "unknown_data_1": chance.name(),
                "unknown_data_2": chance.sentence(),
                "unknown_data_3": chance.word().toUpperCase(),
                "unknown_data_4": chance.domain(),
                "unknown_data_5": chance.domain()
            });
            await ravenSession.advanced.patch(`StakePoolTests/${randomId}`, 'relays', randomRelays());
            await ravenSession.saveChanges();
 
            resolve();
        })
        promiseArr.push(promiseHolder);
        
    }
}


(async () => {
    new Promise(function(){
        // This contains a while loop, so will run until script is stopped.
        query();
    });
    new Promise(function(){
        populate();
    });
})();

```

**Reply by ayende** [reply]
The reason for this behavior is that you are seeing a side benefit of RavenDB. See:
![image](https://user-images.githubusercontent.com/116915/142758594-b994eeac-8e4c-489c-b6c4-c07ae1994e3c.png)

Basically, the output of this query is ~140KB or so. When you run it without having writes in the middle, RavenDB is able to realize that you are running the same query without any changes, so instead of issuing the query, you can utilize the client query cache, significantly improving your perofmrance.

When you have writes of any kind, we have to re-execute the query and send all the results back, this is why you are seeing the different performance numbers.

With `durationInMs: -1,` - it means that it is cached
With `durationInMs: 617,` it means that it was fetched from the server.

The size of your documents is in the 7KB range, so queries on thousands of them will send MBs over the network in many cases, which is costly.

Basically, the issue is whatever RavenDB can use the query cache or not, which depends on whatever you are using caching.
I talk about this in detail here: https://ravendb.net/learn/inside-ravendb-book/reader/4.0/4-deep-dive-into-the-ravendb-client-api#caching

**Reply by CyberCyclone** [reply]
Thanks for that. I was going to let you know that I noticed that the server was hitting a different node so I'm actually seeing latency from the DB node in the wrong location.

I tried again but made sure that I was executing the script on a server in the same location as the DB node and the differences are very small from the patch writes and none batch writes.

**Comment by CyberCyclone** [comment, marked as answer]
In summary, the issue turned out to be the client doing read requests to a DB node in a different datacenter. Due to the size of the data, the network latency was increased due to not being in the same datacenter (it was selecting a different node compared to using the Primary node).

By using the queries 'statistics' option I was able to see that the DB was in a different location. I was able to run the above but in the same datacenter and there was next to no difference in the query times.

Thanks @ayende for helping narrow down the issue.
